File: aidecamp-app/webapi/TurnManager/TurnManager.cs
using Microsoft.SemanticKernel;
using OpenTelemetry.Exporter;
using System.ComponentModel;
public class TurnManagerPlugin
{
    private static readonly Random random = new Random();

    [KernelFunction("calculateOutcome")]
    [Description("Calculate the outcome of the wargame scenario, evaluating modifiers for the firing and target toy soldiers.")]
    [return: Description("The outcome of the wargame scenario.")]
    public ScenarioOutcome CalculateOutcome([Description("The wargame scenario")] ScenarioModel scenario) 
    {
        Console.WriteLine ("Calculating outcome of the wargame scenario");
        Console.WriteLine ("Firing: " + scenario.Firing.Pose + " " + scenario.Firing.Weapon);
        Console.WriteLine ("Target: " + scenario.Target.Pose + " " + scenario.Target.Weapon);
        Console.WriteLine ("Distance: " + scenario.Distance.Value + " " + scenario.Distance.Unit);

        int firingModifier = CalculateFiringModifier(scenario.Firing.Weapon);
        int targetModifier = CalculateTargetModifier(scenario.Target.Pose);
        int distanceModifier = CalculateDistanceModifier(scenario.Distance.Value);

        int rolledDice = random.Next(1, 20);
        int totalModifiers = firingModifier + targetModifier + distanceModifier;
        bool hitOrMiss = rolledDice > totalModifiers;

        return new ScenarioOutcome
        {
            FiringModifier = firingModifier,
            TargetModifier = targetModifier,
            DistanceModifier = distanceModifier,
            RolledDice = rolledDice,
            HitOrMiss = hitOrMiss
        };
    }

    private int CalculateFiringModifier(Weapon weapon)
    {
        return weapon switch
        {
            Weapon.rifle => 1,
            Weapon.machine_gun => 2,
            Weapon.SMG => 3,
            Weapon.pistol => 5,
            _ => 0
        };
    }

    private int CalculateTargetModifier(Pose pose)
    {
        return pose switch
        {
            Pose.crouched => 3,
            Pose.prone => 6,
            Pose.standing => 0,
            _ => 0
        };
    }

    private int CalculateDistanceModifier(int distance)
    {
        return distance >= 70 ? 3 : 0;
    }
}

File: aidecamp-a2a/webapi/turn_manager.py
"""Turn manager for calculating wargame outcomes."""

import random
from typing import Annotated

from semantic_kernel.functions import kernel_function

from models import ScenarioModel, ScenarioOutcome, Weapon, Pose


class TurnManagerPlugin:
    """Plugin for calculating wargame scenario outcomes."""

    def __init__(self):
        self._random = random.Random()

    @kernel_function(
        name="calculate_outcome",
        description="Calculate the outcome of the wargame scenario, evaluating modifiers for the firing and target toy soldiers.",
    )
    def calculate_outcome(
        self, scenario: Annotated[ScenarioModel, "The wargame scenario"]
    ) -> Annotated[ScenarioOutcome, "The outcome of the wargame scenario"]:
        """Calculate the outcome of a wargame scenario."""
        print("Calculating outcome of the wargame scenario")
        print(f"Firing: {scenario.firing.pose} {scenario.firing.weapon}")
        print(f"Target: {scenario.target.pose} {scenario.target.weapon}")
        print(f"Distance: {scenario.distance.value} {scenario.distance.unit}")

        firing_modifier = self._calculate_firing_modifier(scenario.firing.weapon)
        target_modifier = self._calculate_target_modifier(scenario.target.pose)
        distance_modifier = self._calculate_distance_modifier(scenario.distance.value)

        rolled_dice = self._random.randint(1, 20)
        total_modifiers = firing_modifier + target_modifier + distance_modifier
        hit_or_miss = rolled_dice > total_modifiers

        return ScenarioOutcome(
            firing_modifier=firing_modifier,
            target_modifier=target_modifier,
            distance_modifier=distance_modifier,
            rolled_dice=rolled_dice,
            hit_or_miss=hit_or_miss,
        )

    def _calculate_firing_modifier(self, weapon: Weapon) -> int:
        """Calculate the firing modifier based on weapon type."""
        weapon_modifiers = {
            Weapon.RIFLE: 1,
            Weapon.MACHINE_GUN: 2,
            Weapon.SMG: 3,
            Weapon.PISTOL: 5,
        }
        return weapon_modifiers.get(weapon, 0)

    def _calculate_target_modifier(self, pose: Pose) -> int:
        """Calculate the target modifier based on pose."""
        pose_modifiers = {
            Pose.CROUCHED: 3,
            Pose.PRONE: 6,
            Pose.STANDING: 0,
        }
        return pose_modifiers.get(pose, 0)

    def _calculate_distance_modifier(self, distance: int) -> int:
        """Calculate the distance modifier based on distance in cm."""
        return 3 if distance >= 70 else 0